# Fidelity Program - Data Analysis
---------------

<img src="images/ecommerce.png" alt="E-Commerce" style="width:500px; height:320px;"></img>

*Disclaimer: The following context is completely fictitious, the company, the context, the CEO and the business questions. They are all based uppon the recomendation project at [sejaumdatascientist blog](https://sejaumdatascientist.com/como-criar-um-programa-de-fidelidade-para-empresa/).* 

----------------------------


## Introduction
--------------------------

A **Data Science Project** has the objective of create businesses solutions throught insights and predictive models generated by a great Exploratory Data Analysis. The Data Science Project is divided in 7 steps for the problem resolutions' roadmap of Data Science, these being:

1. The Business Question
2. The Business Domain
3. Data Collection
4. Data Cleaning
5. Data Exploration
6. Model Building
7. Data Interpretation

Fullfilling these steps you will create a solution for the business problem. Thereby, this fictitious challenge was created to refine the skills in this 7-steps roadmap. The challenge is an imaginary business problem for a company, as much as the company itself, to simulate a real context.

## Challenge Context
-----------------------------

The **All in One Place** company sells second-line products of several brands at a lower price, through an e-commerce, that is, it is a Multibrand Outlet company.

In a little over a year of operation, the marketing team noticed that some customers in its database buy more expensive products with high frequency and consequently contribute a lot to the company's revenue.

Based on this discovery, the marketing team will launch a loyalty program for the best customers, called **Insiders**. However, the team does not have technical knowledge in data analysis to select program participants. Thus, the marketing team asked the data team for a technical data analysis solution to select eligible customers for the program.

## The Challenge
---

You, as a member of the Data Scientist team, must determine who the customers are eligible to participate in **Insiders**. With this list, the Marketing team will carry out personalized and exclusive strategies and actions to the group, in order to increase the frequency of sales and purchases.

As a project's result, it is expected a list of eligible customers for the program **Insiders** as much as a report responding the following questions:

* **Who are the eligible clients to participate in the Insiders program?**
* **How many clients will participate at the program?**
* **What are the main characteristics of these customers?** 
* **What is the percentage of revenue contribution, coming from Insiders?** 
* **What is this group's revenue expectation for the coming months?**
* **What are the conditions for a person to be eligible for Insiders?**
* **What are the conditions for a person to be removed from Insiders?**
* **What is the guarantee that the Insiders program is better than the rest of the base?**
* **What actions can the marketing team take to increase revenue?**

## The Data
---------------

The dataset is available at the Kaggle platform as [E-Commerce Data](https://www.kaggle.com/carrie1/ecommerce-data).

Each line represents a sales transaction, which took place between the period of November 2016 and December 2017.

The dataset includes the following information:

### Data Description

 **`Invoice Number`**: unique identifier for an invoice. If starts with a "C" it means it was cancelled.\
 **`Stock Code Product`**: item code.\
 **`Description Product`**: item name.\
 **`Quantity`**: The quantity of each item purchased per transaction.\
 **`Invoice Date`**: The day the transaction occurred.\
 **`Unit Price`**: Product price per unit.\
 **`Customer ID`**: unique customer identifier.\
 **`Country`**: The name of the country the customer resides in.\

## Dependencies
--------------

In [1]:
# Sistem and Operational System
import os
import sys

# Data Analysis and Manipulation
import pandas as pd
import numpy as np

# Data Visualization
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import seaborn as sns
sns.set_style("darkgrid")

# GEOJSON
import geojson

# Statistical Models
import statsmodels.api as sm
from statsmodels.sandbox.regression.predstd import wls_prediction_std
import scipy as sp

# Date Time
from datetime import datetime
from datetime import date
import time
from tqdm import tqdm

# Pickle
import joblib

# First, get the path
path = os.getcwd().replace("notebooks","")
path

'C:\\Users\\Avell\\Trabalho\\portfolios\\fidelity_program\\'

## Data Collection
-----------------

In [4]:
data = pd.read_csv(f"{path}data\\data.csv", sep=",")

In [5]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

## Data Cleaning
-----------------------------

Let's if there are null values and the data types of the features.

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [7]:
dtypes = data.dtypes.value_counts().to_frame().rename(columns={0:'dtype'})
dtypes

dtype
object       5
float64      2
int64        1

So we have 8 features, which 5 are strings, 2 are floats and 1 is integer. We also know that 
apparently there are null values. So let's see.

In [8]:
is_null = data.isnull().sum()*100/data.shape[0]
is_null = is_null.to_frame().rename(columns={0:'percent'}).round(2)
is_null

percent
InvoiceNo       0.00
StockCode       0.00
Description     0.27
Quantity        0.00
InvoiceDate     0.00
UnitPrice       0.00
CustomerID     24.93
Country         0.00

So we must deal with **`customer ID`** and **`Description`**. So I will drop the rows with null values, because we need the customers' ID to find them. 

In [9]:
data.dropna(inplace=True)
data.reset_index(drop=True, inplace=True)
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
406824    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
406825    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
406826    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
406827    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
406828    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
406824  12/9/2011 12:50       0.85     12680.0          France  
406825  12/9/2011 12:50       2.10     12680.0          France  
406826  12/9/2011 12:50       4.15     12680.0          France  
406827  12/9/2011 12:50       4.15     12680.0          France  
406828  12/9/2011 12:50       4.95     12680.0          France  

[406829 rows x 8 columns]

Now, let's check some data description

In [10]:
data.describe(include='all')

InvoiceNo StockCode                         Description       Quantity  \
count     406829    406829                              406829  406829.000000   
unique     22190      3684                                3896            NaN   
top       576339    85123A  WHITE HANGING HEART T-LIGHT HOLDER            NaN   
freq         542      2077                                2070            NaN   
mean         NaN       NaN                                 NaN      12.061303   
std          NaN       NaN                                 NaN     248.693370   
min          NaN       NaN                                 NaN  -80995.000000   
25%          NaN       NaN                                 NaN       2.000000   
50%          NaN       NaN                                 NaN       5.000000   
75%          NaN       NaN                                 NaN      12.000000   
max          NaN       NaN                                 NaN   80995.000000   

             InvoiceDate      UnitPrice     CustomerID         Country  
count             406829  406829.000000  406829.000000          406829  
unique             20460            NaN            NaN              37  
top     11/14/2011 15:27            NaN            NaN  United Kingdom  
freq                 543            NaN            NaN          361878  
mean                 NaN       3.460471   15287.690570             NaN  
std                  NaN      69.315162    1713.600303             NaN  
min                  NaN       0.000000   12346.000000             NaN  
25%                  NaN       1.250000   13953.000000             NaN  
50%                  NaN       1.950000   15152.000000             NaN  
75%                  NaN       3.750000   16791.000000             NaN  
max                  NaN   38970.000000   18287.000000             NaN

So, untill now we infer that:

* There are cancelled shops
* There negative values in quantity feature, which doesn't make sense.
* There are 22190 unique invoices.
* There are 3684 unique products.
* There are 3896 unique products names.
* Clearly there are outliers in the quantity feature.
* We have costumers from 37 countries
* And Invoice Date should be a date and not an object.

Let's eliminate the rows of cancelled shops, beacuse it doesn't make any sense.

In [11]:
data= data[data.InvoiceNo.str.contains('C', na=False)==False]
data.reset_index(drop=True, inplace=True)

Now we will eliminate the rows where the **`Quantity`** are below zero

In [12]:
data = data[data.Quantity>0]
data.reset_index(drop=True, inplace=True)

Clearly we have outliers, let's remove them by limiting the quantiles

In [13]:
q1 = data.Quantity.quantile(0.01)
q3 = data.Quantity.quantile(0.99)

high = q3 + 1.5*(q3-q1)
low = q1 - 1.5*(q3-q1)
low = 0 if low <0 else low
high

298.5

So, any quantities above 298 will be transformed into 298

In [14]:
data = data[data.Quantity < int(high)]

Now we will transform the **`InvoiceDate`** as a date data type

In [15]:
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])

In [16]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
397919    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
397920    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
397921    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
397922    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
397923    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                    ...        ...         ...             ...  
397919 2011-12-09 12:50:00       0.85     12680.0          France  
397920 2011-12-09 12:50:00       2.10     12680.0          France  
397921 2011-12-09 12:50:00       4.15     12680.0          France  
397922 2011-12-09 12:50:00       4.15     12680.0          France  
397923 2011-12-09 12:50:00       4.95     12680.0          France  

[396961 rows x 8 columns]

Now, we will infer some new columns for this dataset.

* We can infer a total price feature as the product of quantity and unit price

In [17]:
data['Price'] = data['Quantity']*data['UnitPrice']

In [18]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
397919    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
397920    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
397921    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
397922    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
397923    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  Price  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  15.30  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  20.34  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  22.00  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  20.34  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  20.34  
...                    ...        ...         ...             ...    ...  
397919 2011-12-09 12:50:00       0.85     12680.0          France  10.20  
397920 2011-12-09 12:50:00       2.10     12680.0          France  12.60  
397921 2011-12-09 12:50:00       4.15     12680.0          France  16.60  
397922 2011-12-09 12:50:00       4.15     12680.0          France  16.60  
397923 2011-12-09 12:50:00       4.95     12680.0          France  14.85  

[396961 rows x 9 columns]

## Exploratory Data Analysis

In [55]:
data.describe()

InvoiceNo       Quantity      UnitPrice     CustomerID  \
count  396961.000000  396961.000000  396961.000000  396961.000000   
mean   560619.813460      11.138303       3.120282   15292.998569   
std     13108.652057      21.301452      22.123287    1712.512469   
min    536365.000000       1.000000       0.000000   12347.000000   
25%    549234.000000       2.000000       1.250000   13969.000000   
50%    561893.000000       5.000000       1.950000   15157.000000   
75%    572093.000000      12.000000       3.750000   16794.000000   
max    581587.000000     291.000000    8142.750000   18287.000000   

               Price  
count  396961.000000  
mean       19.980129  
std        77.932122  
min         0.000000  
25%         4.680000  
50%        11.800000  
75%        19.800000  
max     38970.000000

In [58]:
data_test = data[data.Quantity > 27]
data_test.describe()

InvoiceNo      Quantity     UnitPrice    CustomerID         Price
count   24693.000000  24693.000000  24693.000000  24693.000000  24693.000000
mean   560018.165796     73.075649      1.492128  15044.431418    104.568442
std     12778.651239     50.346870      4.376613   1706.910127    279.520619
min    536367.000000     28.000000      0.000000  12347.000000      0.000000
25%    549007.000000     40.000000      0.420000  13694.000000     27.840000
50%    561073.000000     48.000000      1.060000  14712.000000     69.120000
75%    570701.000000     96.000000      1.850000  16546.000000    133.200000
max    581584.000000    291.000000    649.500000  18287.000000  38970.000000

In [59]:
px.box(data_test, y='UnitPrice')

In [60]:
data_test[data_test.UnitPrice==649.5]

InvoiceNo StockCode                     Description  Quantity  \
155035     556444     22502  PICNIC BASKET WICKER 60 PIECES        60   

                InvoiceDate  UnitPrice  CustomerID         Country    Price  
155035  2011-06-10 15:28:00      649.5     15098.0  United Kingdom  38970.0

Now that we cleaned a little bit our data, before we get to the next step, we must understand some metrics called...

### RFM - Recency, Frequency, and Monetary

These metrics are important indicators of the customer's behavior and we will see which each of them means and why they are important.

* **`Recency`**: Is the last time when a customer engaged with a product. Time difference between today and the last purchase

* **`Frequency`**: The total number of transactions or the average time between transactions.

* **`Monetary`**: Total or average transactions value.

So we can say that:

* The lower the recency, the lower the difference between today and the last purchase and more responsive the customer is.

* The higher the frequency, more the customer buys and more satisfied they are.

* Monetary metric segments the customers that spends more, from those that are low-level buyers.

In addition, frequency and monetary affects the metric known as the customer's lifetime value, and the recency says how engaged the customer is with the products.

Soooo, let's get them

In [19]:
today = pd.to_datetime("today")
today
rfm = data.groupby('CustomerID').agg(
    Recency = ('InvoiceDate',lambda date: (today- date.max()).days),
    Frequency = ('InvoiceNo',lambda number: number.nunique()),
    Monetary = ('Price',lambda price: price.sum())
)

In [21]:
rfm.reset_index(level=0, inplace=True)

In [22]:
rfm

CustomerID  Recency  Frequency  Monetary
0        12347.0     3535          7   4310.00
1        12348.0     3608          4   1797.24
2        12349.0     3551          1   1757.55
3        12350.0     3843          1    334.40
4        12352.0     3569          8   2506.04
...          ...      ...        ...       ...
4316     18280.0     3810          1    180.60
4317     18281.0     3713          1     80.82
4318     18282.0     3540          2    178.05
4319     18283.0     3536         16   2094.88
4320     18287.0     3576          3   1837.28

[4321 rows x 4 columns]

In [23]:
rfm.to_csv(f"{path}data\\rfm.csv", sep=',',index=False)

In [24]:
cleaned = data.copy()
cleaned.to_csv(f"{path}data\\cleaned_data.csv",sep=",",index=False)

In [25]:
data = pd.read_csv(f"{path}data\\cleaned_data.csv", sep=",")
rfm = pd.read_csv(f"{path}data\\rfm.csv", sep=",")

In [26]:
rfm

CustomerID  Recency  Frequency  Monetary
0        12347.0     3535          7   4310.00
1        12348.0     3608          4   1797.24
2        12349.0     3551          1   1757.55
3        12350.0     3843          1    334.40
4        12352.0     3569          8   2506.04
...          ...      ...        ...       ...
4316     18280.0     3810          1    180.60
4317     18281.0     3713          1     80.82
4318     18282.0     3540          2    178.05
4319     18283.0     3536         16   2094.88
4320     18287.0     3576          3   1837.28

[4321 rows x 4 columns]

Now we will segment our clients by creating some scores for **Recency**, **Frequency** and **Monetary** features. So we will create some ranges of these features.

In [44]:
rfm.describe()

CustomerID      Recency    Frequency       Monetary
count   4321.000000  4321.000000  4321.000000    4321.000000
mean   15298.030780  3625.218005     4.248554    1835.531542
std     1721.098446   100.023584     7.637460    6749.646682
min    12347.000000  3533.000000     1.000000       2.900000
25%    13812.000000  3550.000000     1.000000     305.280000
50%    15297.000000  3583.000000     2.000000     665.780000
75%    16777.000000  3675.000000     5.000000    1625.970000
max    18287.000000  3906.000000   210.000000  250934.860000

In [46]:
px.box(rfm, y='Recency')

In [47]:
px.box(rfm, y='Frequency')

In [49]:
rfm_test = rfm[rfm.Frequency>12]

In [50]:
rfm_test.describe()

CustomerID      Recency   Frequency       Monetary
count    238.000000   238.000000  238.000000     238.000000
mean   15334.285714  3544.915966   24.726891   13964.320084
std     1735.457345    27.850135   22.490809   25029.065575
min    12415.000000  3533.000000   13.000000    1296.440000
25%    13873.250000  3535.000000   15.000000    4657.635000
50%    15280.500000  3538.000000   18.000000    6746.950000
75%    16775.250000  3546.500000   25.000000   12051.677500
max    18283.000000  3905.000000  210.000000  250934.860000

In [51]:
px.box(rfm, y='Monetary')

In [52]:
rfm_mon = rfm[rfm.Monetary>3600]

In [53]:
rfm_mon.describe()

CustomerID      Recency   Frequency       Monetary
count    435.000000   435.000000  435.000000     435.000000
mean   14942.289655  3556.294253   16.537931   10529.213218
std     1806.790901    41.355276   18.860153   19067.301559
min    12347.000000  3533.000000    1.000000    3603.210000
25%    13204.000000  3536.000000    8.000000    4282.430000
50%    14800.000000  3543.000000   12.000000    5605.970000
75%    16581.000000  3556.000000   19.000000    8939.565000
max    18229.000000  3905.000000  210.000000  250934.860000